### PR TITLE
Webpack & 

### PR DESCRIPTION
**Pull request by nickbalestra**
- moved to webpack:
  1.  using single webpack.config.js 
  2. webpack-dev-server and hrm for developing
  3. css/sass added to pipeline
- added features:
  1. persistent saving of queries
- cleaned dir structure

### CODE REVIEW

File: example/server.js [hunk 1]
```diff
@@ -3,21 +3,28 @@
 const express = require('express')
 const falcorExpress = require('falcor-express')
 const Router = require('falcor-router')
-const falcorPostman = require('./index.js')
+const falcorPostman = require('./../falcor-postman.js')
 const app = express()
+const isDevelopment = process.env.NODE_ENV !== 'production'
+const port = process.env.PORT ? process.env.PORT : 3000
```
Comment (mattiaerre) [inline example/server.js]: I like to declare variables close to usage; shall we move it down again?

File: example/server.js [hunk 2]
```diff
@@ -3,21 +3,28 @@
 const express = require('express')
 const falcorExpress = require('falcor-express')
 const Router = require('falcor-router')
-const falcorPostman = require('./index.js')
+const falcorPostman = require('./../falcor-postman.js')
```
Comment (mattiaerre) [inline example/server.js:6]: so now the middleware is called falcor-postman? I liked index.js tbf

File: .gitignore [hunk 3]
```diff
@@ -2,5 +2,5 @@
 coverage
 jspm_packages
 node_modules
-public
-*.log
\ No newline at end of file
+dist
```
Comment (mattiaerre) [inline .gitignore:5]: lovely; for my info is more common `dist` or `lib`? I guess it depends on what you are building 😅

File: example/server.js [hunk 4]
```diff
@@ -3,21 +3,28 @@
...
+  app.use(falcorPostman({
+    // # middlewarePath
+    // Optional: path used to serve the falcor-postman app, default: '/falcor-postman'
+    // middlewarePath: '/falcor-postman',
 
-  // # falcorPath
-  // Optional: falcor model path, default: localhost/model.json
-  // falcorPath: '/model.json',
+    // # falcorPath
+    // Optional: falcor model path, default: localhost/model.json
```
Comment (mattiaerre) [inline example/server.js]: the default value should only be /model.json surrounded by single quotes right?

File: middleware/index.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import express from 'express'
+import path from 'path'
+
+module.exports = ({middlewarePath='/falcor-postman', app}) => {
+  app.use(express.static(path.join(__dirname, '/../dist')))
+
+  return (req, res, next) => {
+    if (req.url === middlewarePath) {
+      res.status(200)
+      res.send(path.join(__dirname, '/../dist/fp.html'));
```
Comment (mattiaerre) [inline middleware/index.js]: let me try to understand, now the middleware serves a single html file compiled by `webpackServeBundle.js`? if yes why this file is in the `example` folder?

File: package.json [hunk 6]
```diff
@@ -19,50 +19,56 @@
...
   "scripts": {
-    "bundle": "jspm bundle-sfx index.js ./public/js/bundle-sfx.js -m",
-    "coverage": "istanbul cover _mocha -x app/model.js --  --compilers js:babel-core/register --recursive -r jsdom-global/register",
-    "example": "npm run bundle && PORT=3001 nodemon server.js",
-    "start": "npm run bundle && node server.js",
-    "test": "mocha --compilers js:babel-core/register --recursive -r jsdom-global/register"
+    "coverage": "istanbul cover _mocha -x src/model.js --  --compilers js:babel-core/register --recursive -r jsdom-global/register",
+    "start": "node example/server.js",
+    "test": "mocha --compilers js:babel-core/register --recursive -r jsdom-global/register",
+    "build": "rimraf dist && NODE_ENV=production webpack --progress --profile --colors && NODE_ENV=production",
```
Comment (mattiaerre) [inline package.json:25]: what is `rimraf`?

File: package.json [hunk 7]
```diff
@@ -19,50 +19,56 @@
...
+    "production": "npm run build && NODE_ENV=production node example/server.js"
```
Comment (mattiaerre) [inline package.json:26]: this command runs the example using the production code right? not in dev mode to be precise

File: package.json [hunk 8]
```diff
@@ -101,7 +107,8 @@
   "babel": {
     "presets": [
       "es2015",
-      "react"
+      "react",
+      "stage-1"
```
Comment (mattiaerre) [inline package.json]: curiosity; what do we have w/ stage-1?

File: src/app.js [hunk 9]
```diff
@@ -9,23 +12,32 @@ class App extends React.Component {
         button: 'get from query',
       },
       query: '["metrosById", [72], ["name"]]',
```
Comment (mattiaerre) [inline src/app.js:14]: maybe we shouldn't propose this query at the very beginning I think

File: src/app.js [hunk 10]
```diff
@@ -9,23 +12,32 @@ class App extends React.Component {
...
 
-  handleOnClick() {
+  handleOnClick(){
+    const queries = this.state.queries
+    const query = this.state.query
+    if (queries.length === 0 || queries[queries.length - 1] !== query) {
+      const updatedQueries = queries.concat(query)
+      this.setState({queries:updatedQueries })
+      Lockr.set('queries', updatedQueries)
+    }
```
Comment (mattiaerre) [inline src/app.js:40]: basically every time you "press the button" you add the query in the stack right?

File: src/app.js [hunk 11]
```diff
@@ -43,12 +55,27 @@ class App extends React.Component {
...
   render() {
+    const queryHistory = this.state.queries.map((query, i) =>
+      <li key={i} onClick={this.updateQuery.bind(null, query)}>
+        { query }
+      </li>
+    )
+
     return (
       <div className="App">
+        <ul>{queryHistory.reverse()}</ul>
```
Comment (mattiaerre) [inline src/app.js:66]: I ❤️ this!

File: src/static/css/_codeMirror.scss [hunk 12]
```diff
@@ -0,0 +1,336 @@
+.CodeMirror {
```
Comment (mattiaerre) [inline src/static/css/_codeMirror.scss:1]: wow, a pure CSS based editor? nice!

File: src/main.js [hunk 13]
```diff
@@ -2,10 +2,12 @@ import React from 'react'
...
+import 'normalize.css'
+import './static/css/styles.scss'
 
 const props = {
   model,
   falcorPath: '/model.json',
 }
 
-ReactDOM.render(<App {...props} />, document.getElementById('app'))
+ReactDOM.render(<App {...props} />, document.getElementById('main'))
```
Comment (mattiaerre) [inline src/main.js]: no more `app` but `main`? why 😄 ?

File: test/app/app.spec.js [hunk 14]
```diff
@@ -3,11 +3,19 @@ import { shallow, mount } from 'enzyme'
 import chai from 'chai'
 import chaiEnzyme from 'chai-enzyme'
 import sinon from 'sinon'
-import App from '../../app/app'
+import App from './../../src/app'
```
Comment (mattiaerre) [inline test/app/app.spec.js:6]: much better!

**Review by mattiaerre — COMMENTED** [review]
over to you @nickbalestra really impressive work ... let's have a chat about it tomorrow at some point 😉

File: example/server.js (hunk 2 above)
Comment (nickbalestra) [reply example/server.js:6]: As that will be entry point, is common practice to name it with the name of the project. (i.e.:, lodash -> lodash.js, ...).

File: example/server.js (hunk 1 above)
Comment (nickbalestra) [reply example/server.js]: Yes agree, as const/let  are block scoped and don't hoist. I grouped them here while refactoring, will move them closer to usage.

File: example/server.js (hunk 4 above)
Comment (nickbalestra) [reply example/server.js]: yes

File: .gitignore (hunk 3 above)
Comment (nickbalestra) [reply .gitignore:5]: - `dist/`: distribution, it usually contains the compiled software.
- `lib/`: contains the package's dependencies

File: middleware/index.js (hunk 5 above)
Comment (nickbalestra) [reply middleware/index.js]: Thats only a development workflow. 
'webpackServeBundle' allows to run the example in dev mode (`isDevelopment`) serving in that case the bundle via `webpack-dev-middleware` and `webpack-hot-middleware`, thats why is in there.

File: package.json (hunk 6 above)
Comment (nickbalestra) [reply package.json:25]: `rm -rf` for node

File: package.json (hunk 7 above)
Comment (nickbalestra) [reply package.json:26]: Yup. 
- Webpack will use the `build` configuration to create the optimised production bundle and assets.
- the example server will serve those assets without hot-reload enabled

File: package.json (hunk 8 above)
Comment (nickbalestra) [reply package.json]: Its a leftover to be removed. I wanted to use Class properties to make the class definitions for the react component leaner, but enzyme didn't like it so i had to roll it back for the moment :(

File: src/app.js (hunk 9 above)
Comment (nickbalestra) [reply src/app.js:14]: Totally agree

File: src/app.js (hunk 10 above)
Comment (nickbalestra) [reply src/app.js:40]: if the query is not exactly the same as the last you one you just added, yes.

File: src/main.js (hunk 13 above)
Comment (nickbalestra) [reply src/main.js]: too much elm lately... thats why 😄

**Comment by nickbalestra** [comment]
I'm in 🇨🇭 till wed

File: example/server.js (hunk 4 above)
Comment (mattiaerre) [reply example/server.js]: awesome, can you update the description then?

`Optional: falcor model path, default: localhost/model.json` ==> `Optional: falcor model path, default: 'model.json'`

File: middleware/index.js (hunk 5 above)
Comment (mattiaerre) [reply middleware/index.js]: good, but `fp.html` will be **the** production html correct? in that case can we rename it into `falcor-postman.html`?

File: src/app.js (hunk 9 above)
Comment (mattiaerre) [reply src/app.js:14]: let's have a chat today at some point re what it is still outstanding in terms of dependencies that need to be injected

File: src/main.js (hunk 13 above)
Comment (mattiaerre) [reply src/main.js]: alright 😉 ... still prefer app 😜

**Comment by mattiaerre** [comment]
thanks @nickbalestra so maybe a google hangouts today at some point? otherwise I can wait 😉

File: src/app.js (hunk 9 above)
Comment (nickbalestra) [reply src/app.js:14]: Sure, i guess not related to this PR though

File: src/static/css/_codeMirror.scss (hunk 12 above)
Comment (nickbalestra) [reply src/static/css/_codeMirror.scss:1]: not really a pure css editor -> https://codemirror.net/

File: src/main.js (hunk 13 above)
Comment (nickbalestra) [reply src/main.js]: i don't have any preference for this, moved to app

**Comment by nickbalestra** [comment]
@mattiaerre thanks for the review, off to you now.

**Comment by mattiaerre** [comment]
everything looks good to me man ... merging and doing some work on this middleware today as well 😉

**Pull request merged**